Client/SMTPClient.cpp

```diff
@@ -0,0 +1,995 @@
+			throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::SERVER_NOT_RESPONDING);
+		}
+
+		if (result > 0 && FD_ISSET(m_socket, &fdwrite))
+		{
+			result = send(m_socket, &msg_to_send.c_str()[(size_t)index], (int)msg_left, 0);
+			if (result == SOCKET_ERROR || result == 0)
+			{
+				FD_CLR(m_socket, &fdwrite);
+				LOG_ERROR << "Winsock send() error.";
+				throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::WSA_SEND_ERROR);
+			}
+			msg_left	-= result;
+			index		+= result;
+		}
+	}
+
+	FD_CLR(m_socket, &fdwrite);
+
+	return true;
+}
+
+bool	SMTPClientClass::ReceiveData()
+{
+	int		result	= 0;
+	fd_set	fdread;
+	timeval	time;
+
+	time.tv_sec		= m_server_timeout;
+
+	FD_ZERO(&fdread);
+	FD_SET(m_socket, &fdread);
+
+	result = select((int)m_socket + MAX_FILE_DESCRIPTOR, &fdread, NULL, NULL, &time);
+
+	if (result == SOCKET_ERROR)
+	{
+		FD_CLR(m_socket, &fdread);
+		LOG_ERROR << "Winsock select() error.";
+		throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::WSA_SELECT_ERROR);
+	}
+
+	if (result == 0)
+	{		
+		FD_CLR(m_socket, &fdread);
+		LOG_ERROR << "Server is not responding.";
+		throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::SERVER_NOT_RESPONDING);
+	}
+
+	if (FD_ISSET(m_socket, &fdread))
+	{
+		result = recv(m_socket, m_receive_buffer.get(), DEFAULT_BUFFER_SIZE, 0);
+		if (result == SOCKET_ERROR)
+		{
+			FD_CLR(m_socket, &fdread);
+			LOG_ERROR << "Winsock recv() error.";
+			throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::WSA_RECEIVE_ERROR);
+		}
+	}
+
+	FD_CLR(m_socket, &fdread);
+	m_receive_buffer.get()[(size_t)result] = 0;
+
+	if (result == 0)
+	{
+		LOG_ERROR << "Connection with server closed during receive process.";
+		throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::CONNECTION_CLOSED);
+	}
+
+	return true;
+
+}
+
+bool	SMTPClientClass::set_login(const std::string& s)
+{
+	if (s.empty())
+	{
+		LOG_ERROR << "User login is not specified.";
+		throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::STRING_ARGUMENT_EMPTY);
+	}
+	m_login = s;
+
+	return true;
+}
+
+bool	SMTPClientClass::set_port(const std::string& port)
+{
+	if (port.empty())
+	{
+		LOG_ERROR << "Server port is not specified.";
+		throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::STRING_ARGUMENT_EMPTY);
+	}
+	m_port = port;
+
+	return true;
+}
+
+void	SMTPClientClass::set_server_timeout(const int t)
+{
+	m_server_timeout = t;	
+}
+
+bool	SMTPClientClass::set_password(const std::string& s)
+{
+	if (s.empty())
+	{
+		LOG_ERROR << "User password was not specified.";
+		throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::STRING_ARGUMENT_EMPTY);
+	}
+	m_password = s;
+
+	return true;
+}
+
+bool	SMTPClientClass::set_recepient_email(const std::string& s)
+{
+	if (s.empty())
+	{
+		LOG_ERROR << "Recepient email was not specified.";
+		throw SMTPErrorClass(SMTPErrorClass::SMTPErrorEnum::STRING_ARGUMENT_EMPTY);
+	}
+	m_recepient_email = s;
+
+	return true;
+}
+
+bool	SMTPClientClass::set_subject(const std::string& s)
+{
+	if (s.empty())
```

Comment:
I would suggest to allow empty subject. If you are using gmail, you can send email with empty subject